### PR TITLE
Add grad cam for CNN DQN

### PR DESCRIPTION
**Pull request by MrSyee**
- Add grad cam for agent with convolutional layer
- Update README.md

### CODE REVIEW

File: rl_algorithms/common/abstract/agent.py [hunk 1]
```diff
@@ -170,3 +172,80 @@ def _test(self, interim_test: bool = False):
...
+                    result_images = (
+                        result
+                        if result_images is None
+                        else np.vstack([result_images, result])
+                    )
+                # Show action on result image
+                location = (50, 50)
+                font = cv2.FONT_HERSHEY_PLAIN  # hand-writing style font
+                fontScale = 1
+                cv2.putText(
```
Comment (darthegg) [inline rl_algorithms/common/abstract/agent.py:227]: I think there is no need to make new variables for 'font', 'fontScale' and 'location'.
ex) 
`cv2.putText(
    image=result_images,
    text=f"action: {action}",
    org=(50, 50),
    font=cv2.FONT_HERSHEY_PLAIN,
    fontScale=1,
    color=(0, 0, 255),
    thickness=2,
)`

File: rl_algorithms/common/abstract/agent.py [hunk 2]
```diff
@@ -170,3 +172,80 @@ def _test(self, interim_test: bool = False):
...
+                    location,
+                    font,
+                    fontScale,
+                    (0, 0, 255),
+                    2,
+                )
+
+                cv2.imshow("result", result_images)
+                key = cv2.waitKey(0)
+                if key == 27 & 0xFF:
```
Comment (darthegg) [inline rl_algorithms/common/abstract/agent.py:239]: better to add comment that explaining the key.

File: rl_algorithms/common/abstract/agent.py [hunk 3]
```diff
@@ -170,3 +172,80 @@ def _test(self, interim_test: bool = False):
...
+                    break
+
+                state = next_state
+                score += reward
+                step += 1
+
+            print(
+                "[INFO] test %d\tstep: %d\ttotal score: %d" % (i_episode, step, score)
+            )
+            if key == 27 & 0xFF:
```
Comment (darthegg) [inline rl_algorithms/common/abstract/agent.py:250]: better to add comment that explaining the key.

File: rl_algorithms/common/grad_cam.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+class CAMBaseWrapper:
+    """Base Wrapping module for CAM."""
+
+    def __init__(self, model: nn.Module):
+        super(CAMBaseWrapper, self).__init__()
+        self.device = next(model.parameters()).device
+        self.model = model
+        self.handlers = []  # a set of hook function handlers
+
+    def _encode_one_hot(self, ids: torch.Tensor) -> torch.Tensor:
```
Comment (darthegg) [inline rl_algorithms/common/grad_cam.py:28]: no description

File: rl_algorithms/common/grad_cam.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+        # If any candidates are not specified, the hook is registered to all the layers.
+        for name, module in self.model.named_modules():
+            print(name, module)
+            if self.candidate_layers is None or name in self.candidate_layers:
+                self.handlers.append(module.register_forward_hook(forward_hook(name)))
+                self.handlers.append(module.register_backward_hook(backward_hook(name)))
+
+    @staticmethod
+    def _find(pool: OrderedDict, target_layer: str) -> torch.Tensor:
```
Comment (darthegg) [inline rl_algorithms/common/grad_cam.py:99]: no description

File: rl_algorithms/common/grad_cam.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    @staticmethod
+    def _compute_grad_weights(grads: torch.Tensor) -> torch.Tensor:
+        return F.adaptive_avg_pool2d(grads, 1)
+
+    def forward(self, image: np.ndarray) -> torch.Tensor:
+        self.image_shape = image.shape[1:]
+        return super(GradCAM, self).forward(image)
+
+    def generate(self, target_layer: str) -> torch.Tensor:
```
Comment (darthegg) [inline rl_algorithms/common/grad_cam.py:113]: no description

File: rl_algorithms/common/abstract/agent.py (hunk 1 above)
Comment (MrSyee) [reply rl_algorithms/common/abstract/agent.py:226]: OK.

**Pull request merged**